packages/core/contracts/financial-templates/optimistic-rewarder/OptimisticRewarder.sol

```diff
@@ -0,0 +1,429 @@
+        emit UpdateToken(tokenId, msg.sender, data);
+    }
+
+    /**
+     * @notice Requests a redemption for any tokenId. This can be called by anyone.
+     * @dev if called by someone who doesn't own the token, they are effectively gifting their bond to the owner.
+     * @param tokenId the tokenId the redemption is for.
+     * @param cumulativeRedemptions the cumulative token addresses and amounts that this tokenId is eligible for
+     * at the current timestap. cumulative redemptions that are too low should be considered to be valid.
+     */
+    function submitRedemption(uint256 tokenId, RedemptionAmount[] memory cumulativeRedemptions) public nonReentrant {
+        bytes32 redemptionId = _redemptionId(tokenId, cumulativeRedemptions);
+        require(redemptions[redemptionId].expiryTime == 0, "Redemption already exists");
+        require(ownerOf(tokenId) != address(0), "tokenId is invalid");
+
+        uint256 time = getCurrentTime();
+
+        uint256 finalFee = _getStore().computeFinalFee(address(bondToken)).rawValue;
+        bondToken.safeTransferFrom(msg.sender, address(this), finalFee + bond);
+
+        redemptions[redemptionId] = Redemption({ finalFee: finalFee, expiryTime: time + liveness });
+
+        emit Submitted(tokenId, redemptionId, cumulativeRedemptions, time);
+    }
+
+    /**
+     * @notice Disputes a redemption request.
+     * @dev this will cancel a request if the final fee changes or something causes the optimistic oracle proposal to
+     * fail.
+     * @param tokenId the tokenId the redemption is for.
+     * @param cumulativeRedemptions the cumulative redemptions that were provided in the original request.
+     */
+    function dispute(uint256 tokenId, RedemptionAmount[] memory cumulativeRedemptions) public nonReentrant {
+        bytes32 redemptionId = _redemptionId(tokenId, cumulativeRedemptions);
+
+        // This automatically checks that redemptions[redemptionId] != 0.
+        // Check that it has not passed liveness liveness.
+        Redemption storage redemption = redemptions[redemptionId];
+        require(getCurrentTime() < redemption.expiryTime, "redemption expired or nonexistent");
+        if (redemption.finalFee != _getStore().computeFinalFee(address(bondToken)).rawValue) {
+            _cancelRedemption(tokenId, redemptionId);
+            return;
+        } else {
+            SkinnyOptimisticOracleInterface optimisticOracle = _getOptimisticOracle();
+            uint256 totalBond = bond + redemption.finalFee;
+            bondToken.safeApprove(address(optimisticOracle), totalBond * 2);
+            bytes memory ancillaryData =
+                AncillaryData.appendKeyValueBytes32(customAncillaryData, "redemptionId", redemptionId);
+            uint32 requestTimestamp = uint32(redemption.expiryTime - liveness);
+            address proposer = ownerOf(tokenId);
+
+            try
+                optimisticOracle.requestAndProposePriceFor(
```

Comment:
The commenting on the params within this is a bit janky. consider putting the comments on the same line as the variable.

Reply:
Will do.